### ISSUE TITLE
[RFC]: TTS Model Graph Abstraction

### ISSUE THREAD

**Issue by gcanlin** [vllm-project/vllm-omni#2287]
# RFC: TTS Model Graph Abstraction

## Status
- **Authors**: vllm-omni team
- **Created**: 2024
- **Target**: vllm-omni, vllm-ascend

---

## 1. Summary

Move the Graph management logic for TTS models from the Modeling layer to the Runner layer, following the vLLM v1 design pattern and decoupling model code from platform-specific concerns.

---

### Current Issues

```
                    ┌─────────────────────────────────────┐
                    │      Current Implementation (Issues)   │
                    ├─────────────────────────────────────┤
                    │                                     │
  Modeling Layer    │  ┌─────────────────────────────┐   │
  (compute only)     │  │ CUDAGraphDecoderWrapper    │   │  ← Misplaced Graph logic
                    │  │ CUDAGraphAcousticWrapper   │   │
                    │  │ CodePredictorAttention     │   │  ← NPU-specific code
                    │  │   └─ if is_npu: ...        │   │
                    │  └─────────────────────────────┘   │
                    │                                     │
  Runner Layer      │  ┌─────────────────────────────┐   │
                    │  │ talker_mtp: CUDAGraphWrapper│   │  ← Correct pattern
                    │  └─────────────────────────────┘   │
                    │                                     │
                    └─────────────────────────────────────┘
```

**Problems**:
1. `CUDAGraphDecoderWrapper` / `CUDAGraphAcousticTransformerWrapper` manage graphs directly inside the Modeling layer
2. `CodePredictorAttention` contains platform-specific code such as `if is_npu:`
3. Duplicated warmup/capture/replay logic

---

### 3.1 Target Architecture

```
┌─────────────────────────────────────────────────────────────────────────┐
│                              Runner Layer                                │
│                                                                          │
│   ┌──────────────────────────────────────────────────────────────────┐  │
│   │                    GraphWrapper (vLLM/vllm-ascend)               │  │
│   │  ┌─────────────────────┐       ┌─────────────────────┐          │  │
│   │  │  CUDAGraphWrapper   │       │   ACLGraphWrapper   │          │  │
│   │  │  (GPU)              │       │   (NPU)             │          │  │
│   │  └─────────────────────┘       └─────────────────────┘          │  │
│   │              │                           │                       │  │
│   │              └───────────┬───────────────┘                       │  │
│   │                          │                                       │  │
│   │                set_forward_context()                             │  │
│   │                          │                                       │  │
│   └──────────────────────────┼───────────────────────────────────────┘  │
│                              │                                          │
│   ┌──────────────────────────┼───────────────────────────────────────┐  │
│   │                     Model Runner                                 │  │
│   │                          │                                       │  │
│   │    ┌─────────────────────┼─────────────────────┐                │  │
│   │    │                     │                     │                │  │
│   │    ▼                     ▼                     ▼                │  │
│   │ GPUARModelRunner   GPUGenerationRunner   NPUARModelRunner       │  │
│   │    │                     │                     │                │  │
│   │    │ _wrap_components()  │ _wrap_components()  │                │  │
│   │    │ capture_model()     │ capture_model()     │                │  │
│   │    │                     │                     │                │  │
│   └────┼─────────────────────┼─────────────────────┼────────────────┘  │
│        │                     │                     │                    │
├────────┼─────────────────────┼─────────────────────┼────────────────────┤
│        │                     │                     │                    │
│        │              Model Layer (Pure Compute)   │                    │
│        │                     │                     │                    │
│        ▼                     ▼                     ▼                    │
│   ┌─────────────────────────────────────────────────────────────────┐  │
│   │                                                                  │  │
│   │  talker_mtp()      decoder.decode()      acoustic.forward()     │  │
│   │       │                   │                      │               │  │
│   │       │                   │                      │               │  │
│   │       └───────────────────┴──────────────────────┘               │  │
│   │                           │                                      │  │
│   │                  CodePredictorAttention                          │  │
│   │                           │                                      │  │
│   │                           ▼                                      │  │
│   │              ┌─────────────────────────┐                        │  │
│   │              │   AttentionBackend      │  ← Platform abstraction │  │
│   │              │  (SDPA / NPU Fusion)    │                        │  │
│   │              └─────────────────────────┘                        │  │
│   │                                                                  │  │
│   │              * The model is unaware of Graphs                    │  │
│   │              * The model is unaware of the runtime platform      │  │
│   │                                                                  │  │
│   └─────────────────────────────────────────────────────────────────┘  │
│                                                                          │
└─────────────────────────────────────────────────────────────────────────┘
```

### 3.2 Class Diagram

```
                          ┌───────────────────────┐
                          │   OmniGPUModelRunner  │
                          └───────────┬───────────┘
                                      │
              ┌───────────────────────┼───────────────────────┐
              │                       │                       │
              ▼                       ▼                       ▼
   ┌──────────────────┐    ┌──────────────────┐    ┌──────────────────┐
   │ GPUARModelRunner │    │GPUGenerationRunner│   │ OmniNPUModelRunner│
   │                  │    │                  │    │ (+ NPUModelRunner)│
   │ + talker_mtp     │    │ + decoder        │    └────────┬─────────┘
   │   (wrapped)      │    │   (wrapped)      │             │
   └──────────────────┘    └──────────────────┘    ┌────────┴─────────┐
                                                   │                  │
                                                   ▼                  ▼
                                        ┌──────────────┐   ┌──────────────┐
                                        │NPUARModelRunner│ │NPUGenRunner  │
                                        └──────────────┘   └──────────────┘


   Graph Wrapper Selection:
   ┌────────────────────────────────────────────────────────────────────┐
   │                                                                    │
   │   GPU Platform                        NPU Platform                 │
   │   ┌────────────────────┐             ┌────────────────────┐       │
   │   │  CUDAGraphWrapper  │             │   ACLGraphWrapper  │       │
   │   │  (vllm)            │             │   (vllm-ascend)    │       │
   │   └────────────────────┘             └────────────────────┘       │
   │                                                                    │
   └────────────────────────────────────────────────────────────────────┘


   Attention Backend Selection:
   ┌────────────────────────────────────────────────────────────────────┐
   │                      CodePredictorAttention                        │
   │                              │                                     │
   │              get_code_predictor_attn_backend()                     │
   │                              │                                     │
   │              ┌───────────────┴───────────────┐                     │
   │              │                               │                     │
   │              ▼                               ▼                     │
   │   ┌──────────────────┐            ┌──────────────────┐            │
   │   │ SDPABackend      │            │ NPUFusionBackend │            │
   │   │ (GPU/CPU)        │            │ (NPU)            │            │
   │   │                  │            │                  │            │
   │   │ F.sdpa()         │            │ npu_fusion_attn()│            │
   │   └──────────────────┘            └──────────────────┘            │
   │                                                                    │
   └────────────────────────────────────────────────────────────────────┘
```

---

### 4.1 Attention Backend Interface

```python
# vllm_omni/attention/backends/code_predictor.py

from abc import ABC, abstractmethod
from dataclasses import dataclass
import torch

@dataclass
class CodePredictorAttnConfig:
    """Configuration for Code Predictor attention."""
    num_heads: int
    num_kv_heads: int
    head_dim: int
    scale: float
    max_seq_len: int = 2048  # For NPU mask pre-allocation


class CodePredictorAttnBackend(ABC):
    """Abstract backend for Code Predictor self-attention."""

    @staticmethod
    @abstractmethod
    def get_name() -> str:
        """Return backend name for logging."""
        raise NotImplementedError

    @classmethod
    @abstractmethod
    def is_available(cls) -> bool:
        """Check if this backend is available on current platform."""
        raise NotImplementedError

    @abstractmethod
    def forward(
        self,
        query: torch.Tensor,      # [B, num_heads, S, head_dim]
        key: torch.Tensor,        # [B, num_kv_heads, S, head_dim]
        value: torch.Tensor,      # [B, num_kv_heads, S, head_dim]
    ) -> torch.Tensor:            # [B, num_heads, S, head_dim]
        """Compute causal self-attention."""
        raise NotImplementedError
```

### 4.2 Backend Implementations

```python
# vllm_omni/attention/backends/sdpa.py

class SDPACodePredictorBackend(CodePredictorAttnBackend):
    """Standard SDPA backend for GPU/CPU."""

    def __init__(self, config: CodePredictorAttnConfig):
        self.config = config
        self._use_gqa = config.num_heads != config.num_kv_heads

    @staticmethod
    def get_name() -> str:
        return "sdpa"

    @classmethod
    def is_available(cls) -> bool:
        return hasattr(torch.nn.functional, 'scaled_dot_product_attention')

    def forward(self, query, key, value) -> torch.Tensor:
        return F.scaled_dot_product_attention(
            query, key, value,
            scale=self.config.scale,
            is_causal=True,
            enable_gqa=self._use_gqa,
        )
```

```python
# vllm_ascend/attention/backends/npu_fusion_code_predictor.py

class NPUFusionCodePredictorBackend(CodePredictorAttnBackend):
    """NPU Fusion Attention backend (PR #2695 optimization)."""

    def __init__(self, config: CodePredictorAttnConfig):
        self.config = config
        # Pre-allocate causal mask for NPU fusion attention
        self._causal_mask = torch.triu(
            torch.ones(config.max_seq_len, config.max_seq_len, dtype=torch.bool),
            diagonal=1,
        )
        self._use_gqa = config.num_heads != config.num_kv_heads

    @staticmethod
    def get_name() -> str:
        return "npu_fusion"

    @classmethod
    def is_available(cls) -> bool:
        try:
            import torch_npu
            return hasattr(torch_npu, 'npu_fusion_attention')
        except ImportError:
            return False

    def forward(self, query, key, value) -> torch.Tensor:
        import torch_npu
        seq_len = query.shape[2]

        # GQA expansion
        if self._use_gqa:
            repeat = self.config.num_heads // self.config.num_kv_heads
            key = key.repeat_interleave(repeat, dim=1)
            value = value.repeat_interleave(repeat, dim=1)

        # NPU layout: [B, N, S, D] -> [B, S, N, D]
        q = query.transpose(1, 2).contiguous()
        k = key.transpose(1, 2).contiguous()
        v = value.transpose(1, 2).contiguous()

        out = torch_npu.npu_fusion_attention(
            q, k, v,
            atten_mask=self._causal_mask[:seq_len, :seq_len],
            scale=self.config.scale,
            pre_tockens=self.config.max_seq_len,
            next_tockens=0,
        )[0]

        return out.transpose(1, 2)
```

### 4.3 Backend Selector

```python
# vllm_omni/attention/backends/selector.py

from vllm_omni.platforms import current_omni_platform

_BACKEND_REGISTRY: dict[str, type[CodePredictorAttnBackend]] = {}

def register_backend(name: str):
    """Decorator to register a backend."""
    def decorator(cls):
        _BACKEND_REGISTRY[name] = cls
        return cls
    return decorator

def get_code_predictor_attn_backend(
    config: CodePredictorAttnConfig
) -> CodePredictorAttnBackend:
    """Get the best available attention backend."""

    # Priority: platform-specific > generic
    if current_omni_platform.is_npu():
        try:
            from vllm_ascend.attention.backends.npu_fusion_code_predictor import (
                NPUFusionCodePredictorBackend
            )
            if NPUFusionCodePredictorBackend.is_available():
                return NPUFusionCodePredictorBackend(config)
        except ImportError:
            pass

    # Fallback to SDPA
    from vllm_omni.attention.backends.sdpa import SDPACodePredictorBackend
    return SDPACodePredictorBackend(config)
```

### 4.4 Runner Graph Wrapper Integration

```python
# vllm_omni/worker/gpu_generation_model_runner.py

class GPUGenerationModelRunner(OmniGPUModelRunner):

    def load_model(self):
        super().load_model()
        self._wrap_generation_components()

    def _wrap_generation_components(self):
        """Wrap decoder/acoustic with CUDAGraphWrapper."""
        if not self._should_use_cuda_graph():
            return

        from vllm.compilation.cuda_graph import CUDAGraphWrapper

        component = self._get_graphable_component()
        if component is None:
            return

        # Wrap the forward method
        original_forward = component.forward
        component.forward = CUDAGraphWrapper(
            original_forward,
            self.vllm_config,
            runtime_mode=CUDAGraphMode.FULL,
        )
        self._graph_wrapped_component = component

    def _get_graphable_component(self):
        """Get the component to wrap (model-specific)."""
        # Qwen3-TTS
        if hasattr(self.model, 'decoder'):
            return self.model.decoder
        # Voxtral
        if hasattr(self.model, 'acoustic_transformer'):
            return self.model.acoustic_transformer
        return None

    def capture_model(self) -> int:
        result = super().capture_model()
        self._capture_generation_graphs()
        return result

    def _capture_generation_graphs(self):
        """Capture graphs via set_forward_context."""
        if not hasattr(self, '_graph_wrapped_component'):
            return

        from vllm.distributed.parallel_state import graph_capture

        capture_sizes = self._compute_capture_sizes()
        with torch.inference_mode(), graph_capture(device=self.device):
            for size in capture_sizes:
                dummy = self._create_dummy_input(size)
                batch_desc = BatchDescriptor(num_tokens=size)

                with set_forward_context(
                    None, self.vllm_config,
                    cudagraph_runtime_mode=CUDAGraphMode.FULL,
                    batch_descriptor=batch_desc,
                ):
                    self._graph_wrapped_component.forward(dummy)
```

```python
# vllm_omni/platforms/npu/worker/npu_generation_model_runner.py

class NPUGenerationModelRunner(OmniNPUModelRunner):

    def _wrap_generation_components(self):
        """Wrap with ACLGraphWrapper for NPU."""
        if not self._should_use_acl_graph():
            return

        from vllm_ascend.compilation.acl_graph import ACLGraphWrapper

        component = self._get_graphable_component()
        if component is None:
            return

        component.forward = ACLGraphWrapper(
            component.forward,
            self.vllm_config,
            runtime_mode=CUDAGraphMode.FULL,
        )
        self._graph_wrapped_component = component
```

---

### Phase 1: Add Attention Backend (Non-breaking)

```
vllm_omni/attention/backends/
├── __init__.py
├── code_predictor.py          # Abstract interface
├── sdpa.py                    # SDPA implementation
└── selector.py                # Backend selector

vllm_ascend/attention/backends/
└── npu_fusion_code_predictor.py  # NPU implementation
```

### Phase 2: Migrate CodePredictorAttention

```diff
# vllm_omni/model_executor/models/common/qwen3_code_predictor.py

class CodePredictorAttention(nn.Module):
    def __init__(self, config, ...):
        ...
+       from vllm_omni.attention.backends.selector import (
+           get_code_predictor_attn_backend,
+           CodePredictorAttnConfig,
+       )
+       self._attn_backend = get_code_predictor_attn_backend(
+           CodePredictorAttnConfig(
+               num_heads=self.num_heads,
+               num_kv_heads=self.num_kv_heads,
+               head_dim=self.head_dim,
+               scale=self.scaling,
+           )
+       )

    def forward(self, hidden_states, position_embeddings):
        ...
-       attn_out = F.scaled_dot_product_attention(
-           q, k, v, scale=self.scaling, is_causal=True, enable_gqa=...
-       )
+       attn_out = self._attn_backend.forward(q, k, v)
        ...
```

### Phase 3: Move Graph Logic to Runner

```diff
# Delete these files:
- vllm_omni/model_executor/models/qwen3_tts/cuda_graph_decoder_wrapper.py
- vllm_omni/model_executor/models/voxtral_tts/cuda_graph_acoustic_transformer_wrapper.py

# Simplify model code:
# vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code2wav.py
class Qwen3TTSCode2Wav:
    def __init__(self, ...):
        self.decoder = Qwen3TTSTokenizerV2Decoder(...)
-       self.cuda_graph_decoder = CUDAGraphDecoderWrapper(self.decoder, ...)

    def forward(self, codes):
-       return self.cuda_graph_decoder.decode(codes)
+       return self.decoder.decode(codes)  # Runner handles graph
```

---

## 6. Summary

| Before | After |
|--------|-------|
| Graph logic in Modeling | Graph logic in Runner |
| Platform code in Model | Platform code in Backend |
| Custom graph wrappers | Reuse vLLM `CUDAGraphWrapper`/`ACLGraphWrapper` |
| Duplicated warmup/capture | Unified via `set_forward_context` |

**Key Benefits**:
1. Model code stays pure (compute only)
2. Platform differences abstracted via Backend
3. Reuses vLLM infrastructure
4. PR #2695 integrates cleanly via NPU Backend

**Comment by hsliuustc0106** [comment]
@linyueqian PTAL

**Comment by linyueqian** [comment]
Thanks for the writeup. The duplication is real.

The 4-way split looks clean. A few thoughts:

**Workload adapter interface:** The graphable workloads across models differ more than the RFC suggests:
- Qwen3-TTS decoder captures a single forward pass with frame-length buckets and output waveform slicing
- Qwen3-TTS code predictor captures a re-prefill transformer forward per batch-size bucket, with pre-allocated projection buffers and torch.compile layered on top
- VoxtralTTS captures a multi-step flow (greedy argmax + 8-step Euler ODE + quantization) with noise buffers that must be refilled before each replay
- Fish Speech fast AR uses torch.compile with dynamic shapes instead of CUDA graphs, with no bucket capture at all

The adapter API needs to handle cases like VoxtralTTS's pre-replay buffer mutation and the code predictor's compile+graph layering. Worth sketching the interface against at least the Qwen3 decoder, VoxtralTTS, and code predictor before committing to the contract.

**torch.compile:** The code predictor already layers torch.compile on top of CUDA graphs, and Fish Speech uses compile instead of graphs entirely. I'd suggest the adapter accept an optional compiled callable from the start rather than retrofitting later.

**Testing:** A mock backend for CI without GPU would help validate the runner logic independently.

**Comment by hsliuustc0106** [comment]
cc @xuechendi @tjtanaa

**Comment by xuechendi** [comment]
XPU does not support XPU graph before PT2.11, will follow after pt upgrade

**Comment by ZeldaHuang** [comment]
The talker stage in Qwen3-Omni should also use this cuda graph abstraction

**Comment by hsliuustc0106** [comment]
> The talker stage in Qwen3-Omni should also use this cuda graph abstraction

this RFC deserves further discussion, please submit an issue to this week's meeting

**Comment by hsliuustc0106** [comment]
@gcanlin still working on it?
